lib/verify-stream.js

```diff
@@ -0,0 +1,120 @@
+/*global module*/
+const base64url = require('base64url');
+const DataStream = require('./data-stream');
+const jwa = require('jwa');
+const Stream = require('stream');
+const toString = require('./tostring');
+const util = require('util');
+const JWS_REGEX = /^[a-zA-Z0-9\-_]+?\.[a-zA-Z0-9\-_]+?\.([a-zA-Z0-9\-_]+)?$/;
+
+function isObject(thing) {
+  return Object.prototype.toString.call(thing) === '[object Object]';
+}
+
+function safeJsonParse(thing) {
+  if (isObject(thing))
+    return thing;
+  try { return JSON.parse(thing); }
+  catch (e) { return undefined; }
+}
+
+function headerFromJWS(jwsSig) {
+  const encodedHeader = jwsSig.split('.', 1)[0];
+  return safeJsonParse(base64url.decode(encodedHeader, 'binary'));
+}
+
+function securedInputFromJWS(jwsSig) {
+  return jwsSig.split('.', 2).join('.');
+}
+
+function algoFromJWS(jwsSig) {
+  var err;
+  const header = headerFromJWS(jwsSig);
+  if (typeof header != 'object') {
+    err = new Error("Invalid token: no header in signature '" + jwsSig + "'");
+    err.code = "MISSING_HEADER";
+    err.signature = jwsSig;
+    throw err;
+  }
+  if (!header.alg) {
+    err = new Error("Missing `alg` field in header for signature '"+ jwsSig +"'");
+    err.code = "MISSING_ALGORITHM";
+    err.header = header;
+    err.signature = jwsSig;
+    throw err;
+  }
+  return header.alg;
+}
+
+function signatureFromJWS(jwsSig) {
+  return jwsSig.split('.')[2];
+}
+
+function payloadFromJWS(jwsSig) {
+  const payload = jwsSig.split('.')[1];
+  return base64url.decode(payload, 'binary');
+}
+
+function isValidJws(string) {
+  return JWS_REGEX.test(string) && !!headerFromJWS(string);
```

Comment:
This is the only actual [code change](https://github.com/brianloveswords/node-jws/pull/16/files#diff-168726dbe96b3ce427e7fedce31bb0bcL87) which I can split out into another pull request if you would like.